### PR TITLE
Multi-worker support in Pytorch Dataset

### LINKED ISSUE #145: Pytorch DataLoader mulit-process support
# Problem Statement

As per Pytorch document

https://pytorch.org/docs/stable/data.html#iterable-style-datasets

> When num_workers > 0, each worker process will have a different copy of the dataset object, so it is often desired to configure each copy independently to avoid having duplicate data returned from the workers. [get_worker_info()](https://pytorch.org/docs/stable/data.html#torch.utils.data.get_worker_info), when called in a worker process, returns information about the worker.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #145

### CODE REVIEW

**Comment by eddyxu** [comment]
Screenshot 2022-09-08 at 12 06 17 PM

File: python/lance/__init__.py [hunk 1]
```diff
@@ -19,20 +19,19 @@
...
     register_extension_types()
 
-__all__ = ["dataset", "write_table", "scanner"]
+__all__ = ["dataset", "write_table", "scanner", "LanceFileFormat"]
 
 
 def dataset(
     uri: str,
-) -> ds.Dataset:
+) -> ds.FileSystemDataset:
```
Comment (changhiskhan) [inline python/lance/__init__.py:34]: why not leave it more generic? do we use FileSystemDataset-specific APIs?

File: python/lance/pytorch/data.py [hunk 2]
```diff
@@ -52,26 +55,47 @@ class LanceDataset(IterableDataset):
...
+    def _setup_dataset(self):
+        """Lazy loading dataset in different process."""
+        if self._files is not None:
+            return self._files
+
+        self._fs, _ = pyarrow.fs.FileSystem.from_uri(self.root)
+        self._files = dataset(self.root).files
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info:
+            # Split the work using at the files level for now.
```
Comment (changhiskhan) [inline python/lance/pytorch/data.py:83]: so just to check my understanding, this is what's forcing us to split into many smaller files right now right? Theoretically if we just have some num_rows based parallelism we could have good analytics performance and good training scan performance?

**Review by changhiskhan — APPROVED** [review]
can we make it so the `max_rows_per_file` and `max_rows_per_group` is passed in explicitly via commandline options in `parse_pet.py`? the default behavior should be the same as before so analytics performance stays the same if we need to re-run benchmark numbers

File: python/lance/__init__.py (hunk 1 above)
Comment (eddyxu) [reply python/lance/__init__.py:34]: FileSystemDataset has this `dataset.files` attributes.

File: python/lance/pytorch/data.py (hunk 2 above)
Comment (eddyxu) [reply python/lance/pytorch/data.py:83]: Yes.

**Pull request merged**